### PR TITLE
Add metrics

### PR DESCRIPTION
**Pull request by urbanserj**
JIRA: https://jira.mesosphere.com/browse/DCOS_OSS-4756

### CODE REVIEW

File: src/lashup_kv.erl [hunk 1]
```diff
@@ -470,3 +534,72 @@ kv2map(New, [Old | _]) ->
...
+    {registry, lashup},
+    {name, kv_backend_read_seconds},
+    {labels, [table]},
+    {duration_unit, seconds},
+    {help, "The time spent reading data from KV backend."}
+  ]),
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, kv_backend_write_seconds},
+    {labels, [table]},
```
Comment (GoelDeepak) [inline src/lashup_kv.erl]: It would be good to also capture information about `Key` as well

File: src/lashup_gm_mc.erl [hunk 2]
```diff
@@ -194,10 +209,46 @@ forward_packet(MulticastPacket, Tree) ->
...
+  ]),
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, mc_send_packets_seconds},
+    {duration_unit, seconds},
+    {help, "The time spent sending multicast packets."}
+  ]),
+  prometheus_counter:new([
+    {registry, lashup},
+    {name, mc_send_bytes},
```
Comment (GoelDeepak) [inline src/lashup_gm_mc.erl]: would it be possible to have `mc_send_packets`, a count?

File: src/lashup_gm_mc.erl [hunk 3]
```diff
@@ -194,10 +209,46 @@ forward_packet(MulticastPacket, Tree) ->
...
+    {help, "Total number of multicast packets sent in bytes."}
+  ]),
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, mc_receive_bytes},
```
Comment (GoelDeepak) [inline src/lashup_gm_mc.erl]: Similarly, `mc_receive_packets`

File: src/lashup_kv_sync_rx_fsm.erl [hunk 4]
```diff
@@ -70,3 +80,15 @@ handle_disconnect({'DOWN', _MonitorRef, _Type, _Object, noconnection}) ->
...
+%%%===================================================================
+%%% Metrics functions
+%%%===================================================================
+
+-spec(init_metrics() -> ok).
+init_metrics() ->
+    prometheus_summary:new([
+        {registry, lashup},
+        {name, aae_rx_bytes},
+        {help, "The size of AAE RX messages received by node in bytes."}
```
[the remark this answers is not in the document]
Comment (urbanserj) [reply src/lashup_kv_sync_rx_fsm.erl]: The `summary` metric type is actually the total sum of all observed values, the total number of events that have been observed, and φ-quantiles of observed events [1]. The last one is not implemented in `prometheus.erl` library. So there are actually two metrics here: `aae_rx_bytes_sum` and `aae_rx_bytes_count`. The semantics seems fine for `_seconds` metrics, but doesn't look good for `_bytes`. I'm considering to have two counters in this case: `aae_rx_bytes_total` and `aae_rx_messages_total`. Please let me know what do you think.

[1] https://prometheus.io/docs/concepts/metric_types/#summary

File: src/lashup_kv.erl (hunk 1 above)
Comment (urbanserj) [reply src/lashup_kv.erl]: Prometheus docs recommend not to add such kind of labels [1]. We can add key-specific metrics in place where we use them in dcos-net.

[1] https://prometheus.io/docs/practices/naming/#labels

File: src/lashup_gm_mc.erl (hunk 2 above)
Comment (urbanserj) [reply src/lashup_gm_mc.erl]: There is `mc_send_packets_seconds_count`. It's a part of `mc_send_packets_seconds` summary metric.

File: src/lashup_gm_mc.erl (hunk 3 above)
Comment (urbanserj) [reply src/lashup_gm_mc.erl]: Please see https://github.com/dcos/lashup/pull/49#discussion_r264088340

File: src/lashup_kv_sync_tx_fsm.erl [hunk 5]
```diff
@@ -132,3 +135,24 @@ handle_disconnect({'DOWN', _MonitorRef, _Type, _Object, noconnection}) ->
...
+
+%%%===================================================================
+%%% Metrics functions
+%%%===================================================================
+
+-spec(init_metrics() -> ok).
+init_metrics() ->
+    prometheus_summary:new([
+        {registry, lashup},
+        {name, aae_tx_bytes},
```
[the remark this answers is not in the document]
Comment (urbanserj) [reply src/lashup_kv_sync_tx_fsm.erl]: Please see https://github.com/dcos/lashup/pull/49#discussion_r264088340.

File: src/lashup_gm_mc.erl [hunk 6]
```diff
@@ -5,7 +5,8 @@
 %% API
 -export([
   start_link/0,
-  multicast/2
+  multicast/2,
```
Comment (Fabs) [inline src/lashup_gm_mc.erl]: nit: the title of the commit has a typo: s/milticast/multicast

Not important, but maybe you wanna fix when you squash.

File: src/lashup_gm_mc.erl [hunk 7]
```diff
@@ -100,15 +101,26 @@ code_change(_OldVsn, State, _Extra) ->
...
 %%  -Buffer messages if my neighbors / active view are flapping too much
 %%  -Experiment with adding the tree that I build to the messages.
 %%   Although this will make messages larger, it might be more efficient (I also don't know how much larger)
 %%   The tree representation could also be significantly reduced in size rather than the map
 %%   Which has a bunch of extraneous metadata
 
--spec(handle_do_original_multicast(topic(), payload()) -> ok).
-handle_do_original_multicast(Topic, Payload) ->
+-spec(do_original_multicast(topic(), payload()) -> ok).
+do_original_multicast(Topic, Payload) ->
```
Comment (Fabs) [inline src/lashup_gm_mc.erl]: would it also be feasible to call this method `original_multicast`?

File: src/lashup_kv_sync_rx_fsm.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (Fabs) [reply src/lashup_kv_sync_rx_fsm.erl]: Also, will `aae` makes sense for the customers of this metric? Should we also document it a little more somewhere? (design part of our docs, readme on this repo).

File: src/lashup_kv.erl (hunk 1 above)
Comment (Fabs) [reply src/lashup_kv.erl]: What about the cardinality of tables in lashup? I think I only saw two tables on a quick test I did, but I wonder how many you have on a cluster, and if the size of the cluster matters for the number of tables.

File: src/lashup_gm.erl [hunk 8]
```diff
@@ -445,7 +434,42 @@ persist(Member, _State) ->
...
+  ]),
+  prometheus_gauge:new([
+    {registry, lashup},
+    {name, gm_subscribers},
+    {help, "The number of global membership subscribers."}
+  ]),
+  prometheus_counter:new([
+    {registry, lashup},
+    {name, gm_updates_total},
+    {help, "Total number of global membership table updates."}
```
Comment (ichernetsky) [inline src/lashup_gm.erl]: nit: I guess, it should start with `The` since the other ones do :)

File: src/lashup_gm_mc.erl [hunk 9]
```diff
@@ -191,148 +166,89 @@ do_original_cast(Node, Packet) ->
...
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, mc_send_packets_seconds},
+    {duration_unit, seconds},
+    {help, "The time spent sending multicast packets."}
+  ]),
+  prometheus_counter:new([
+    {registry, lashup},
+    {name, mc_send_bytes_total},
+    {help, "Total number of multicast packets sent in bytes."}
```
Comment (ichernetsky) [inline src/lashup_gm_mc.erl]: ditto :)

File: src/lashup_hyparview_membership.erl [hunk 10]
```diff
@@ -1011,6 +1021,54 @@ handle_recognize_pong(_Pong = #{now := _Now, receiving_node := Node},
...
+  prometheus_gauge:new([
+    {registry, lashup},
+    {name, hyparview_passive_view_size},
+    {help, "The size of HyParView Passive View."}
+  ]),
+  prometheus_counter:new([
+    {registry, lashup},
+    {name, hyparview_cast_messages_total},
+    {labels, [type]},
+    {help, "Total number of HyParView messages sent to other nodes."}
```
Comment (ichernetsky) [inline src/lashup_hyparview_membership.erl]: ditto :)

File: test/lashup_kv_aae_SUITE.erl [hunk 11]
```diff
@@ -26,6 +26,7 @@ init_per_suite(Config) ->
     [{hostname, Hostname} | Config].
 
 end_per_suite(Config) ->
+    application:stop(prometheus),
```
Comment (ichernetsky) [inline test/lashup_kv_aae_SUITE.erl]: Should it be stopped after `lashup`?

File: src/lashup_kv.erl (hunk 1 above)
Comment (urbanserj) [reply src/lashup_kv.erl]: There are only two mnesia tables in lashup: `kv` stores `riak_dt_map`s, `nclock` stores metadata for AAE.

File: src/lashup_gm.erl (hunk 8 above)
Comment (urbanserj) [reply src/lashup_gm.erl]: It was intentional. I checked several Erlang projects and almost all of them have "Total number of" for counter metrics, but "The time/bytes/size/lenght/etc" for other metric types. I've check it once again. 70% projects on github (all languages, mostly go) have prometheus metrics with "Total number of" help string and only 30% have metrics with help that starts with "The total number".

File: src/lashup_kv_sync_rx_fsm.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (urbanserj) [reply src/lashup_kv_sync_rx_fsm.erl]: AAE is a well-known abbreviation. I believe anyone who works with such kind of databases know what it means.

[the remark this answers is not in the document]
Comment (urbanserj) [reply src/lashup_kv_sync_rx_fsm.erl]: `aae_rx_bytes_total` and `aae_rx_messages_total` have been added.

File: src/lashup_kv_sync_tx_fsm.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (urbanserj) [reply src/lashup_kv_sync_tx_fsm.erl]: I've added `aae_tx_bytes_total` and `aae_tx_messages_total`.

File: src/lashup_gm_mc.erl (hunk 3 above)
Comment (urbanserj) [reply src/lashup_gm_mc.erl]: Done, here are new metrics: `mc_receive_bytes_total` and `mc_receive_messages_total`.

File: src/lashup_kv_sync_rx_fsm.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (GoelDeepak) [reply src/lashup_kv_sync_rx_fsm.erl]: 👍

**Pull request merged**